File: utils/src/main/java/cm/aptoide/pt/utils/q/QManager.java
package cm.aptoide.pt.utils.q;

import android.app.ActivityManager;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.util.Base64;
import android.view.WindowManager;

import static cm.aptoide.pt.utils.AptoideUtils.ScreenU;
import static cm.aptoide.pt.utils.AptoideUtils.SystemU;

/**
 * Created by neuro on 12-05-2017.
 */
public class QManager {

  private final GlExtensionsManager glExtensionsManager;
  private final Resources resources;
  private final ActivityManager activityManager;
  private final WindowManager windowManager;
  private Integer minSdk;
  private String cpuAbi;
  private String screenSize;
  private String glEs;
  private Integer densityDpi;
  private String cachedFilters;

  public QManager(SharedPreferences sharedPreferences, Resources resources,
      ActivityManager activityManager, WindowManager windowManager) {
    this.glExtensionsManager = new GlExtensionsManager(sharedPreferences);
    this.resources = resources;
    this.activityManager = activityManager;
    this.windowManager = windowManager;
  }

  public Integer getMinSdk() {
    if (minSdk == null) {
      minSdk = computeMinSdk();
    }
    return minSdk;
  }

  public String getCpuAbi() {
    if (cpuAbi == null) {
      cpuAbi = computeCpuAbi();
    }
    return cpuAbi;
  }

  public String getScreenSize() {
    if (screenSize == null) {
      screenSize = computeScreenSize();
    }
    return screenSize;
  }

  public String getGlEs() {
    if (glEs == null) {
      glEs = computeGlEs();
    }
    return glEs;
  }

  public Integer getDensityDpi() {
    if (densityDpi == null) {
      densityDpi = computeDensityDpi();
    }
    return densityDpi;
  }

  private int computeMinSdk() {
    return SystemU.getSdkVer();
  }

  private String computeScreenSize() {
    return ScreenU.getScreenSize(resources);
  }

  private String computeGlEs() {
    return SystemU.getGlEsVer(activityManager);
  }

  private int computeDensityDpi() {
    return ScreenU.getDensityDpi(windowManager);
  }

  private String computeCpuAbi() {
    return SystemU.getAbis();
  }

  public String getSupportedOpenGlExtensionsManager() {
    return glExtensionsManager.getSupportedExtensions();
  }

  public boolean isSupportedExtensionsDefined() {
    return glExtensionsManager.isSupportedExtensionsDefined();
  }

  public String getFilters(boolean hwSpecsFilter) {
    if (!hwSpecsFilter) {
      return null;
    }

    if (cachedFilters == null) {
      cachedFilters = computeFilters();
    }

    return cachedFilters;
  }

  private String computeFilters() {
    String filters = "maxSdk="
        + getMinSdk()
        + "&maxScreen="
        + getScreenSize()
        + "&maxGles="
        + getGlEs()
        + "&myCPU="
        + getCpuAbi()
        + "&myDensity="
        + getDensityDpi()
        + (getSupportedOpenGlExtensionsManager().equals("") ? ""
        : "&myGLTex=" + getSupportedOpenGlExtensionsManager());

    return Base64.encodeToString(filters.getBytes(), 0)
        .replace("=", "")
        .replace("/", "*")
        .replace("+", "_")
        .replace("\n", "");
  }

  private void invalidate() {
    cachedFilters = null;
  }

  public void setSupportedOpenGLExtensions(String openGLExtensions) {
    glExtensionsManager.setSupportedOpenGLExtensions(openGLExtensions);
    invalidate();
  }
}

File: aptoide-network/src/main/java/cm/aptoide/pt/aptoide_network/q/QManager.kt
package cm.aptoide.pt.aptoide_network.q

import android.app.ActivityManager
import android.app.UiModeManager
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Build
import android.os.Build.VERSION
import android.util.Base64

class QManager(
  private val resources: Resources,
  private val activityManager: ActivityManager,
  private val uiModeManager: UiModeManager
) {
  private var minSdk: Int? = null
  private var cpuAbi: String? = null
  private var screenSize: String? = null
  private var glEs: String? = null
  private var densityDpi: Int? = null
  private var cachedFilters: String? = null

  fun getFilters(hwSpecsFilter: Boolean): String? {
    if (!hwSpecsFilter) return null
    cachedFilters ?: run { cachedFilters = computeFilters() }
    return cachedFilters
  }

  private fun computeFilters(): String {
    val filters = "maxSdk=${getMinSdk()}" +
      "&maxScreen=${getScreenSize()}" +
      "&maxGles=${getGlesVer()}" +
      "&myCPU=${getCpuAbi()}" +
      "&leanback=${hasLeanback()}" +
      "&myDensity=${getDensityDpi()}"

    return Base64.encodeToString(filters.toByteArray(), 0)
      .replace("=", "")
      .replace("/", "*")
      .replace("+", "_")
      .replace("\n", "")
  }

  private fun getMinSdk(): Int? {
    minSdk ?: run { minSdk = VERSION.SDK_INT }
    return minSdk
  }

  private fun getScreenSize(): String? {
    screenSize ?: run { screenSize = computeScreenSize() }
    return screenSize
  }

  private fun computeScreenSize(): String =
    when (getScreenSizeInt()) {
      Configuration.SCREENLAYOUT_SIZE_UNDEFINED -> "undefined"
      Configuration.SCREENLAYOUT_SIZE_SMALL -> "small"
      Configuration.SCREENLAYOUT_SIZE_NORMAL -> "normal"
      Configuration.SCREENLAYOUT_SIZE_LARGE -> "large"
      Configuration.SCREENLAYOUT_SIZE_XLARGE -> "xlarge"
      else -> "unknown"
    }

  private fun getScreenSizeInt(): Int =
    resources.configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK

  private fun getGlesVer(): String? {
    glEs ?: run { glEs = activityManager.deviceConfigurationInfo.glEsVersion }
    return glEs
  }

  private fun getCpuAbi(): String? {
    cpuAbi ?: run { cpuAbi = Build.SUPPORTED_ABIS.joinToString { "," } }
    return cpuAbi
  }

  private fun hasLeanback(): String {
    return if (uiModeManager.currentModeType == Configuration.UI_MODE_TYPE_TELEVISION)
      "1" else "0"
  }

  private fun getDensityDpi(): Int? {
    densityDpi ?: run { densityDpi = computeDensityDpi() }
    return densityDpi
  }

  private fun computeDensityDpi(): Int = resources.displayMetrics.densityDpi
    .let { dpi ->
      when {
        dpi <= 120 -> 120
        dpi <= 160 -> 160
        dpi <= 213 -> 213
        dpi <= 240 -> 240
        dpi <= 320 -> 320
        dpi <= 480 -> 480
        else -> 640
      }
    }
}
